### ISSUE TITLE
Ctrl-C broken – repeating Ctrl-C required on Julia 1.8

### ISSUE THREAD

**Issue by fonsp** [JuliaLang/julia#46635]
Since Julia 1.8, a **single Ctrl-C does not work**. I always need to repeat Ctrl-C multiple times to get the forced SIGINT (with `WARNING: Force throwing a SIGINT`):

https://user-images.githubusercontent.com/6933510/188428251-dec8a3da-7410-4ebf-b001-8de77457e14b.mov

I found this to be the case for anything that I interrupt, not just `sleep`. This means that, to exit Pluto, Jupyter, a web server or any other long running process, users will now need to **hold down** Ctrl-C, [printing multiple WARNINGs](https://user-images.githubusercontent.com/6933510/188687702-daf3e8f4-7bbc-447e-afde-7c9c03caca32.png), which makes Julia look unstable.

```
Julia Version 1.8.0
Commit 5544a0fab76 (2022-08-17 13:38 UTC)
Platform Info:
  OS: macOS (x86_64-apple-darwin21.4.0)
  CPU: 8 × Apple M1
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-13.0.1 (ORCJIT, westmere)
  Threads: 1 on 8 virtual cores
```
```
Julia Version 1.7.0
Commit 3bf9d17731 (2021-11-30 12:12 UTC)
Platform Info:
  OS: macOS (x86_64-apple-darwin19.5.0)
  CPU: Apple M1
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-12.0.1 (ORCJIT, westmere)
```

Same problem with the MacOS ARM build.

**Comment by pankgeorg** [comment]
I can confirm I have the same issue on Windows 11 on an Intel processor 

![image](https://user-images.githubusercontent.com/8681967/188454205-c4137e0f-89dc-4fb9-8dc6-ffaf5c1b6ca8.png)


```
julia> versioninfo()
Julia Version 1.8.0
Commit 5544a0fab7 (2022-08-17 13:38 UTC)
Platform Info:
  OS: Windows (x86_64-w64-mingw32)
  CPU: 8 × 11th Gen Intel(R) Core(TM) i7-1165G7 @ 2.80GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-13.0.1 (ORCJIT, tigerlake)
  Threads: 1 on 8 virtual cores
```

**Comment by gbaraldi** [comment]
A bisect shows that the first bad commit is https://github.com/JuliaLang/julia/commit/0eaf35a1990bd572a3bf701aa3e8623224b005c5 .
@vtjnash

**Comment by fonsp** [comment]
It looks like it is still possible to interrupt Pkg precompilation with a single Ctrl+C ([video](https://user-images.githubusercontent.com/6933510/191289490-c45f635e-6483-4337-a14b-6d0e3177ac52.mov)), strange...

**Comment by fonsp** [comment]
Narrowing down on the interrupt in Pkg precompilation, I found that `Ctrl+C` still works reliably in this snippet: 

```julia
while true
	sleep(.1)
end
```


<details>
<summary>Details</summary>


```
julia> sleep(100)
^C^C^C^C^C^C^CWARNING: Force throwing a SIGINT
ERROR: InterruptException:

julia> while true
       sleep(100)
       end
^C^C^C^C^C^C^C^CWARNING: Force throwing a SIGINT
ERROR: InterruptException:

julia> while true
       sleep(.1)
       end
^CERROR: InterruptException:
Stacktrace:
 [1] poptask(W::Base.InvasiveLinkedListSynchronized{Task})
   @ Base ./task.jl:921
 [2] wait()
   @ Base ./task.jl:930
 [3] wait(c::Base.GenericCondition{Base.Threads.SpinLock})
   @ Base ./condition.jl:124
 [4] _trywait(t::Timer)
   @ Base ./asyncevent.jl:138
 [5] wait
   @ ./asyncevent.jl:155 [inlined]
 [6] sleep(sec::Float64)
   @ Base ./asyncevent.jl:240
 [7] top-level scope
   @ ./REPL[14]:2

julia> 
```

</details>


This means that Packages that use the "Press Ctrl+C to stop" UI, like an HTTP.jl server, can use this as a **workaround**.

## Example with HTTP.jl:

Instead of:

```julia
server = HTTP.serve!(...)

@info "Server running, press Ctrl+C to interrupt..."
wait(server)
```

You can use:

```julia
server = HTTP.serve!(...)

@info "Server running, press Ctrl+C to interrupt..."
try
	while Base.isopen(server)
		sleep(.1)
	end
catch e
	@info "Closing server..."
	close(server)
	wait(server)
end

```

**Comment by vtjnash** [comment]
The more correct way to read ^C in a UI is exemplified here in runtests.jl:
https://github.com/JuliaLang/julia/blob/276af848511510861f8d0e6c1e69e15e4c646e3e/test/runtests.jl#L221-L232

**Comment by quinnj** [comment]
I don't think it's reasonable to expect HTTP.jl internals to be messing with REPL/Terminals/raw character input. I also don't really understand what changed though and why doing Ctrl + C doesn't throw an InteruptException like it used to? Maybe we just need to provide an `interuptible_wait()` function in Base that people can call if they want a way to reliably allow and catch an InteruptException?

**Comment by mind6** [comment]
does a test framework exist for testing keystroke interrupts?

**Comment by AbhimanyuAryan** [comment]
yes same issue MacOS and Windows

**Comment by toollu** [comment]
I don't understand why this is not patched in 1.8.x but put back for 1.9 which is probably months from a release

**Comment by PallHaraldsson** [comment]
Julia 1.9 seems actually close, but I believe there's no solution [PR] yet (though a workaround in Pluto), so I'm afraid it will not make it into 1.9, maybe 1.10, which IS months away. Though if solved likely could be backported to 1.9. I'm guessing the solution isn't too easy (and the behavoir that worked, was never promised, so some tradeoff with keeping it, not simply a bug).
